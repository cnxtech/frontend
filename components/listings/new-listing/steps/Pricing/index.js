import React, { Component } from 'react'
import { Formik, Field } from 'formik'
import MaskedInput from 'react-text-mask'

import Icon from '@emcasa/ui-dom/components/Icon'
import Button from '@emcasa/ui-dom/components/Button'
import Input from '@emcasa/ui-dom/components/Input'
import Row from '@emcasa/ui-dom/components/Row'
import Col from '@emcasa/ui-dom/components/Col'
import View from '@emcasa/ui-dom/components/View'
import Text from '@emcasa/ui-dom/components/Text'
import StaticMap from 'components/listings/new-listing/shared/StaticMap'
import NavButtons from 'components/listings/new-listing/shared/NavButtons'
import {
  currencyInputMask,
  currencyStyle,
  PREFIX,
  THOUSANDS_SEPARATOR_SYMBOL
} from 'utils/text-utils'

class Pricing extends Component {
  constructor(props) {
    super(props)
    this.nextStep = this.nextStep.bind(this)
    this.previousStep = this.previousStep.bind(this)
    this.updateStateFromProps = this.updateStateFromProps.bind(this)
    this.priceSuggestion = this.priceSuggestion.bind(this)
    this.noPriceSuggestion = this.noPriceSuggestion.bind(this)

    this.userPriceInput = null
  }

  state = {
    userPrice: null,
    suggestedPrice: null,
    editingPrice: false
  }

  componentDidMount() {
    this.updateStateFromProps(this.props)
    if (this.userPriceInput !== null) {
      this.userPriceInput.focus()
    }
  }

  componentWillReceiveProps(props) {
    this.updateStateFromProps(props)
  }

  updateStateFromProps(props) {
    const { pricing } = props
    if (pricing) {
      this.setState({
        userPrice: pricing.userPrice,
        suggestedPrice: pricing.suggestedPrice,
        editingPrice: pricing.editingPrice
      })
    }
  }

  parseUserPrice(userPrice) {
    const cleanUserPrice = userPrice.replace(PREFIX, '').replace(THOUSANDS_SEPARATOR_SYMBOL, '')
    return parseInt(cleanUserPrice)
  }

  nextStep() {
    const { navigateTo, updatePricing } = this.props
    const intUserPrice = this.parseUserPrice(this.state.userPrice)
    const intSuggestedPrice = parseInt(this.state.suggestedPrice)
    const newPricing = {
      userPrice: intUserPrice,
      suggestedPrice: intSuggestedPrice,
      editingPrice: this.state.editingPrice
    }
    updatePricing(newPricing)
    navigateTo('services')
  }

  previousStep() {
    const { navigateTo } = this.props
    navigateTo('personal')
  }

  currencyInput(errors, setFieldValue, setFieldTouched) {
    const { userPrice } = this.props.pricing
    return (
      <Field
        name="userPrice"
        validate={this.validateUserPrice}
        render={({form}) => (
          <MaskedInput
            mask={currencyInputMask}
            render={(ref, props) =>
              <Input
                {...props}
                hideLabelView
                error={form.touched.userPrice ? errors.userPrice : null}
                placeholder="R$ 000.000"
                defaultValue={userPrice}
                onChange={(e) => {
                  const { value } = e.target
                  setFieldValue('userPrice', value)
                  setFieldTouched('userPrice')
                  this.setState({userPrice: value})
                }}
                ref={(input) => {
                  this.userPriceInput = input
                  return ref(input)
                }}
              />
            }
          />
        )}
      />
    )
  }

  priceSuggestion(errors, setFieldValue, setFieldTouched) {
    const { pricing } = this.props
    const formattedSuggestedPrice = pricing.suggestedPrice.toLocaleString('pt-BR', currencyStyle)
    return (
      <Col>
        <Text color="grey">Seu imóvel foi avaliado por:</Text>
        <Text fontSize="large" fontWeight="bold" textAlign="center">{formattedSuggestedPrice}</Text>
        <Text color="grey">Recomendamos anunciar por:</Text>
            {this.state.editingPrice ?
              <Col width={[1, 1/2]} mr={4}>
                {this.currencyInput(errors, setFieldValue, setFieldTouched)}
              </Col>
              :
              <>
                <Row justifyContent="center">
                  <Col style={{width: 100}}></Col>
                  <Col>
                    <Text inline fontSize="large" fontWeight="bold" textAlign="center">{formattedSuggestedPrice}</Text>
                  </Col>
                  <Col style={{width: 100}}>
                    <Button onClick={() => this.setState({editingPrice: true})} style={{marginLeft: 20}}>
                      <Icon name="pen" color="dark" />
                    </Button>
                  </Col>
                </Row>
              </>
            }
        <Text color="grey">Não gostou da nossa avaliação? Não tem problema. É só editar o valor do seu imóvel.</Text>
      </Col>
    )
  }

  noPriceSuggestion(errors, setFieldValue, setFieldTouched) {
    return (
      <>
        <Row>
          <Col>
            <Text color="grey">Seu imóvel será avaliado por um de nossos agentes, mas conte pra gente por quanto você gostaria de vender seu imóvel.</Text>
          </Col>
        </Row>
        <Row>
          <Col width={[1, 1/2]} mr={4}>
            {this.currencyInput(errors, setFieldValue, setFieldTouched)}
          </Col>
        </Row>
      </>
    )
  }

  validateUserPrice(value) {
    if (!value || value === 'R$ ') {
      return 'É necessário informar um preço de venda.'
    }
  }

  render() {
    const { pricing, location } = this.props
    let suggestedPrice, userPrice, addressData
    if (pricing && location) {
      suggestedPrice = pricing.suggestedPrice
      userPrice = pricing.userPrice
      addressData = location.addressData
    }
    return (
      <div ref={this.props.hostRef}>
        <Row justifyContent="center">
          <Col width={[1, 1/2]}>
            <Formik
              initialValues={{
                userPrice: userPrice
              }}
              isInitialValid={() => {
                return true
              }}
              render={({isValid, setFieldTouched, setFieldValue, errors}) => (
                <>
                  <View body p={4}>
                    <Text
                      fontSize="large"
                      fontWeight="bold"
                      textAlign="center">
                      Qual o valor do seu imóvel?
                    </Text>
                    <Col>
                      <StaticMap addressData={addressData} />
                    </Col>
                    {suggestedPrice ?
                      this.priceSuggestion(errors, setFieldValue, setFieldTouched)
                    :
                      this.noPriceSuggestion(errors, setFieldValue, setFieldTouched)
                    }
                  </View>
                  <View bottom p={4}>
                    <NavButtons
                      previousStep={this.previousStep}
                      nextStep={this.nextStep}
                      nextEnabled={isValid}
                    />
                  </View>
                </>
              )}
            />
          </Col>
        </Row>
      </div>
    )
  }
}

export default Pricing
