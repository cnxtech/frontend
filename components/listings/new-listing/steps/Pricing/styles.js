import styled from 'styled-components'
import { themeGet } from 'styled-system'

import Col from '@emcasa/ui-dom/components/Col'
import Button from '@emcasa/ui-dom/components/Button'

const UserPriceCol = styled(Col)`
  width: 100px;
`
const EditPriceButton = styled(Button)`
  margin-left: 20px;
`

export {
  UserPriceCol,
  EditPriceButton
}
