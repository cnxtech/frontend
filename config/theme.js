import theme from '@emcasa/ui'
import {fas} from '@fortawesome/pro-solid-svg-icons'
import {fal} from '@fortawesome/pro-light-svg-icons'
import icons from './icons'

export default {
  ...theme,
  icons: {emcasa: icons, default: fas, light: fal}
}
