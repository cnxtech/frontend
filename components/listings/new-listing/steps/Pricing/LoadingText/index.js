import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { PoseGroup } from 'react-pose'
import { FadeLoader } from 'react-spinners'
import theme from '@emcasa/ui'
import Text from '@emcasa/ui-dom/components/Text'
import Row from '@emcasa/ui-dom/components/Row'
import FadeInOut from 'components/shared/Animation/FadeInOut'
import { randomKey } from 'lib/random'

const TIME_PER_SENTENCE = 3000

class LoadingText extends Component {
  constructor(props) {
    super(props)
    this.state = {
      currentSentence: 0
    }
    this.swapSentence = this.swapSentence.bind(this)
    this.sentenceTimer = null
  }

  componentDidMount() {
    this.sentenceTimer = setInterval(this.swapSentence, TIME_PER_SENTENCE)
  }

  componentWillUnmount() {
    clearInterval(this.sentenceTimer)
  }

  swapSentence() {
    const { sentences } = this.props
    let nextSentence = ++this.state.currentSentence
    if (nextSentence >= sentences.length) {
      nextSentence = 0
    }
    this.setState({ currentSentence: nextSentence})
  }

  render() {
    const { sentences } = this.props
    return (
      <>
        <PoseGroup>
          {sentences.map((sentence, index) => {
            if (this.state.currentSentence === index) {
              return (
                <FadeInOut key={randomKey()}>
                  <Row width={1} justifyContent="center">
                    <Text key={index}>{sentence}</Text>
                  </Row>
                </FadeInOut>
              )
            }
          })}
        </PoseGroup>
        <Row mt={4} flexDirection="column" alignItems="center" width={1}>
          <FadeLoader
            width={10}
            height={10}
            margin={2}
            radius={8}
            color={theme.colors.pink}
          />
        </Row>
      </>
    )
  }
}

LoadingText.propTypes = {
  sentences: PropTypes.array.isRequired
}

export default LoadingText
