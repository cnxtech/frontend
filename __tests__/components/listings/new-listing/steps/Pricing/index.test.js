import React from 'react'
import Pricing from 'components/listings/new-listing/steps/Pricing'
import renderer from 'react-test-renderer'
import 'jest-styled-components'
import theme from '@emcasa/ui'
import { ThemeProvider } from 'styled-components'

it('renders an empty state', () => {
  const pricing = {
    suggestedPrice: '1000',
    userPrice: '1100',
    editingPrice: null
  }

  const tree = renderer.create(
    <ThemeProvider theme={theme}>
      <Pricing
        location={{}}
        pricing={pricing}
      />
    </ThemeProvider>).toJSON()
  expect(tree).toMatchSnapshot()
})
