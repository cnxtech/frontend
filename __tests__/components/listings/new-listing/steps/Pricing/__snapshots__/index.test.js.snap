// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders an empty state 1`] = `
.c2 {
  -webkit-font-smoothing: antialiased;
  font-family: FaktSoftPro-Normal;
  color: #3c4858;
  font-size: 24px;
  font-weight: bold;
  text-align: center;
}

.c3 {
  -webkit-font-smoothing: antialiased;
  font-family: FaktSoftPro-Normal;
  color: #979797;
  font-size: 18px;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding: 20px;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c8 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin-top: 20px;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c1 {
  width: 100%;
}

.c9 {
  width: 41.66666666666667%;
}

.c7 {
  outline: none;
  cursor: pointer;
  padding: 0 20px 0 20px;
  border-radius: 4px;
  border: 1px solid;
  border-color: #e0e6ed;
  background-color: #ffffff;
  width: auto;
  height: 40px;
  color: #3c4858;
  font-family: FaktSoftPro-Normal;
  font-size: 18px;
  -webkit-font-smoothing: antialiased;
  border-color: #e0e6ed;
}

.c7 :hover {
  background-color: #ffffff;
}

.c10 {
  outline: none;
  display: block;
  cursor: pointer;
  padding: 0 20px 0 20px;
  border-radius: 4px;
  border: 1px solid;
  border-color: #e0e6ed;
  background-color: #ffffff;
  width: 100%;
  height: 60px;
  color: #3c4858;
  font-family: FaktSoftPro-Normal;
  font-size: 18px;
  -webkit-font-smoothing: antialiased;
  border-color: #e0e6ed;
}

.c10 :hover {
  background-color: #ffffff;
}

.c11 {
  outline: none;
  display: block;
  cursor: pointer;
  padding: 0 20px 0 20px;
  border-radius: 4px;
  border: 1px solid;
  border-color: #f50057;
  background-color: #f50057;
  width: 100%;
  height: 60px;
  color: #ffffff;
  font-family: FaktSoftPro-Normal;
  font-size: 18px;
  -webkit-font-smoothing: antialiased;
  border-color: #e0e6ed;
}

.c11 :hover {
  background-color: #f50057;
}

.c5 {
  width: 100px;
}

.c6 {
  margin-left: 20px;
}

@media screen and (min-width:640px) {
  .c1 {
    width: 50%;
  }
}

<div>
  <div
    className="c0"
  >
    <div
      className="c1"
      width={
        Array [
          1,
          0.5,
        ]
      }
    >
      <p
        className="c2"
        color="dark"
        fontSize="large"
        fontWeight="bold"
      >
        Qual o valor do seu imóvel?
      </p>
      <div
        className=""
      />
      <div
        className=""
      >
        <p
          className="c3"
          color="grey"
          fontSize="default"
        >
          Seu imóvel foi avaliado por:
        </p>
        <p
          className="c2"
          color="dark"
          fontSize="large"
          fontWeight="bold"
        >
          R$1,000
        </p>
        <p
          className="c3"
          color="grey"
          fontSize="default"
        >
          Recomendamos anunciar por:
        </p>
        <div
          className="c4"
        >
          <div
            className="c5 "
          />
          <div
            className=""
          >
            <span
              className="c2"
              color="dark"
              fontSize="large"
              fontWeight="bold"
            >
              R$1,100
            </span>
          </div>
          <div
            className="c5 "
          >
            <button
              className="c6 c7"
              fontSize="default"
              height="medium"
              onClick={[Function]}
              style={
                Object {
                  "marginLeft": 20,
                }
              }
            >
              <div
                className=""
              >
                <svg
                  height={20}
                  version="1.1"
                  viewBox="0 0 512 512"
                  x={0}
                  y={0}
                >
                  <path
                    d="M290.74 93.24l128.02 128.02-277.99 277.99-114.14 12.6C11.35 513.54-1.56 500.62.14 485.34l12.7-114.22 277.9-277.88zm207.2-19.06l-60.11-60.11c-18.75-18.75-49.16-18.75-67.91 0l-56.55 56.55 128.02 128.02 56.55-56.55c18.75-18.76 18.75-49.16 0-67.91z"
                    fill="#3c4858"
                    name="pen"
                    type="default"
                  />
                </svg>
              </div>
            </button>
          </div>
        </div>
        <p
          className="c3"
          color="grey"
          fontSize="default"
        >
          Não gostou da nossa avaliação? Não tem problema. É só editar o valor do seu imóvel.
        </p>
      </div>
      <div
        className="c8"
      >
        <div
          className="c9"
          width={0.4166666666666667}
        >
          <button
            className="c10"
            fontSize="default"
            height="tall"
            onClick={[Function]}
          >
            Voltar
          </button>
        </div>
        <div
          className="c9"
          width={0.4166666666666667}
        >
          <button
            className="c11"
            disabled={false}
            fontSize="default"
            height="tall"
            onClick={[Function]}
          >
            Avançar
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
